File: 22-generate-parentheses/generate-parentheses.java
import java.util.*;

class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> ans = new ArrayList<>();
        backtrack(ans, new StringBuilder(), 0, 0, n);
        return ans;
    }

    private void backtrack(List<String> ans, StringBuilder cur, int open, int close, int n) {
        if (cur.length() == 2 * n) {
            ans.add(cur.toString());
            return;
        }
        if (open < n) {
            cur.append('(');
            backtrack(ans, cur, open + 1, close, n);
            cur.deleteCharAt(cur.length() - 1);
        }
        if (close < open) {
            cur.append(')');
            backtrack(ans, cur, open, close + 1, n);
            cur.deleteCharAt(cur.length() - 1);
        }
    }
}

File: 22-generate-parentheses/generate-parentheses.cpp
class Solution {
public:
    vector<string> ans;
    
    void solve(int& n, int open, string temp ){
        if(temp.size() == 2* n ){
            if(open ==0)
                ans.push_back(temp);
            // cout<<temp<<" ";
            return ;
        }


        if(open<n){
            temp += '(';
            solve(n , open +1, temp);
            temp.pop_back();
        }

        if(open >0){
            temp += ')';
            solve(n, open -1, temp);
            temp.pop_back();
        }

    }


    vector<string> generateParenthesis(int n) {
        solve(n, 0, "");
        return ans;
    }
};